### ISSUE TITLE
allow metadata array of strings to be attached to a survey response

### ISSUE THREAD

**Issue by eric-jahn** [hserv/coordinated-entry#234]
Survey users will use this to attach tags, scores, etc. to responses.

cc: @CTAJavier

**Comment by CTAJavier** [comment]
tl;dr version.  we want to submit a single response for a question composed of a response"Response" and  a value "1"  that would then appear in the data warehouse as 2 columns.  response and response_value

Details below.


currently submitting a response looks like this 
```
{
  "responses": {
    "responses": [
      {
        "questionId": "a4a8baf6-363a-4be8-8610-2cdb6593316f",
        "responseText": "example response",
        "sectionId": "b5b8cdg8-363a-4be8-8610-2cdb6593316f",
        "appId": "app1",
        "effectiveDate": "10-02-2017 03:12:00.000"
      },
      {
        "questionId": "b5b8cdg8-363a-4be8-8610-2cdb6593316f",
        "responseText": "example response",
        "sectionId": "a4a8baf6-363a-4be8-8610-2cdb6593316f",
        "appId": "app2",
        "effectiveDate": "10-02-2017 03:12:00.000"
      }
    ]
  }
}
```

The key here is the "responseText". Whatever is in the responseText is what will be displayed in the data warehouse.   If we wanted to show a value instead of the literal response, currently we could do that by sending the value instead of the literal response, IE instead of sending "Example response" we could send "1". If we do this we lose the ability to see the literal response in the data warehouse.

To make things more complicated, this would only work if each response had a static value ie "Example response" would always ="1"

If we wanted to use logic to calculate our value it would not be supported.
, ie, question "how many years have you been homeless", the client is show a textbox where they enter a number for their response, depending on the number we calculate a value.  For Example, 

- Response = 1,    Value= 1, 
- Response = 2 or 3,    Value = 2, 
- Response = 4,    Value = 3 
- Response = 5  to 10,    Value = 4
- Response = 10- 24,   Value = 5
- Response = 25+,  Value = 10
To be clear these are not choices, we are looking at the number entered by a user and assiging a value based on that response.

The value would be unique to the Survey, the same question could have different values in different surveys. or no value at all. 

If say we submited for  "responseText":{ "example response","1:} we could support both Values.
Data warehouse should then create to columns for this question Response and Response_value on creation of the custom table when the survey is created.

**Comment by eric-jahn** [comment]
Okay, I see two distinct example uses you provided: 
* additionally storing mapped encodings of the literal response and/or 
* storing derived encodings of the literal response.  
Essentially metadata about the literal response you've chosen to initially store.  The number of additional metadata would be unbounded, and you would be responsible for knowing what it means.  You mention encodings, but it really could be anything you wish to pack along with the "responseText" value.  It could be tags, code versions, dates: you name it. 

[As an aside, I'm in unknown territory here, since I'm really only used to systems that store the mapped encoded response and the code set/version.  I've never witnessed storage of the literal response.]

So the responseText JSON element would be changed from:
"responseText": "example response as a code or literal",
to:
"responseText" : [ "example response as a code or literal", "metadata", "metadata", ... ],

So the convention would be that the first member of the response array is the actual responseText value.  Everything after is metadata.  One example could be:

"responseText" : [ "green eye color", "3", "1"],
where it's "responseText" : [ "literal response", "mapped code response", "derived response"],

or without literals:
"responseText" : [ "3", "1"], 
where it's "responseText" : [ "mapped code response", "derived response"],

or with a tag:
"responseText" : [ "3", "1","urgent"], 
where it's "responseText" : [ "mapped code response", "derived response", "tag"],

or all the above:
"responseText" : [ "green eye color", "3", "1","urgent"], 
where it's "responseText" : [ "literal response", "mapped code response", "derived response", "tag"],

But regardless,  the first value will be considered the responseText's "official value".  Just a thought: if we are going to create extra columns anyway in the Big Data warehouse for the additional array metadata items, why don't we just avoid appending the metadata to the responseText, and just make a separate metadata JSON element.  So it would look like this: 

"responseText" : [ "green eye color"], 
"responseMetadata" : ["3", "1","urgent"],

Then there would not need to be a convention about the first array element being the actual responseText value.  But it' up to the client to know what the values mean.  I mean you could store something like :
"responseMetadata" : [{"code":"3"}, {"derived_code":"1"},{"tag":"urgent"}],

or even 

"responseMetadata" : [{"code":"3"}, {"code_version":"v3.201"}, {"derived_code":"1"}, {"tag":"urgent"}],

**Comment by eric-jahn** [comment]
@logicsandeep and @suryayadavalli , can you provide time estimates for each of your portions on this issue? Thanks!

**Comment by logicsandeep** [comment]
Copying our chat conversations here.
@eric-jahn Sorry I had a few questions and that's why I did not add any estimates.

1) Should we show the json as it is in the datawarehouse, or create columns dynamically based on the response_text (json). (edited)
2) Will the question_type be different for these responses ?
If we had to dynamically create new columns in the data warehouse depending on the response_text (json) how do I know which response would have a json and which one would have a full text.

eric-jahn [9:38 AM]
@sandeep.dolia You can just add those questions as github comments, just to keep it all tidy.

Sandeep Dolia [9:38 AM]
Yes  I plan to do that @eric-jahn

eric-jahn [9:38 AM]
Also, that way, Javier can help also
I like the idea of a separate data element called responseMetadata.  It's an array also.
To answer #2, no, the question type won't change.  responseMetadata can be added as an element to any question_type.
@sandeep.dolia Is there a big time difference if, for your question 1), we have to unpack each array item into a separate column dynamically?
But I don't know the answer to 1).

Sandeep Dolia [10:02 AM]
I like the idea of a separate column

eric-jahn [10:03 AM]
What about difficulty/time required?

eric-jahn [10:37 AM]
@sandeep.dolia Javier wants a separate column for each.
@sandeep.dolia So does that give you enough to estimate?

Sandeep Dolia [10:40 AM]
Sure @eric-jahn I'll populate the estimates very soon

eric-jahn [10:42 AM]
Can can you quote two ways?  One is if the responseMetadata is only one value, not an array.  The other quote is if it's an array of any length.

Sandeep Dolia [10:59 AM]
@eric-jahn Why would you expect responseMetadata to be just a value ?  We have response_text to populate the value.

eric-jahn [11:16 AM]
@sandeep.dolia Sorry, I meant "if the responseMetadata is only one metadata value, not an array of metadata values".
We're leaving the responseText value alone...

Sandeep Dolia [11:18 AM]
@eric-jahn This is my understanding, responseMetadata is a new column to the response table and anything that goes inside responseMetadata needs to be in the json format. And even if it has 1 value it needs to be in the json format. (edited)

eric-jahn [12:07 PM]
So the remaining question is, should it stay in JSON format, or lose the formatting (lose the quotes, brackets, ellipses)?

**Comment by eric-jahn** [comment]
> <@U0JPE255G> btw, I got an answer from Javier about the JSON formatting.  It should be unpacked without the JSON formatting.  No quotes, elllipses. <https://hslynk.slack.com/archives/G0JLWRQ1H/p1536858431000100>
<sub>[View in Slack](https://hslynk.slack.com/archives/G0JLWRQ1H/p1536860684000100)</sub>

**Comment by eric-jahn** [comment]
Closing because customer no longer wants this,
